### PR TITLE
Set queue family index to 0 on creating buffer and image during replaying

### PR DESCRIPTION
**Pull request by locke-lunarg**
When the trimming capturing initizes, it needs to deal with buffers and images on ProcessBufferMemory and ProcessImageMemory. But when the buffers and images are created with queue family index not 0, it might happen DeivceLost on QueueWaitIdle or QueueSubmit of ProcessBufferMemory  or ProcessImageMemory.

related: https://github.com/LunarG/Projects/issues/528

### CODE REVIEW

**Comment by locke-lunarg** [comment]
https://github.com/LunarG/Projects/issues/528
For this issue, it always failed during trimming capturing. The reason is `Device_Lost` on `WaitForQueue`. The queue's family index is 1. I figured out if I changed it to 0, the trimming will success. Honestly, I don't exact know why it happens `Device_Lost`, and I'm not sure if it's a good idea to change it to 0.

File: framework/decode/vulkan_replay_consumer_base.cpp [hunk 1]
```diff
@@ -4155,23 +4155,32 @@ VulkanReplayConsumerBase::OverrideCreateBuffer(PFN_vkCreateBuffer
...
     VkResult                              result = VK_SUCCESS;
     VulkanResourceAllocator::ResourceData allocator_data;
-    auto                                  replay_buffer      = pBuffer->GetHandlePointer();
-    auto                                  capture_id         = (*pBuffer->GetPointer());
-    auto                                  replay_create_info = pCreateInfo->GetPointer();
-
+    auto                                  replay_buffer               = pBuffer->GetHandlePointer();
+    auto                                  capture_id                  = (*pBuffer->GetPointer());
+    auto                                  replay_create_info          = pCreateInfo->GetPointer();
+    VkBufferCreateInfo                    modified_create_info        = *replay_create_info;
```
Comment (andrew-lunarg) [inline framework/decode/vulkan_replay_consumer_base.cpp:4161]: We used to have a null check on line 4220 but now we trust this isn't null. Is that safe?

File: framework/decode/vulkan_replay_consumer_base.cpp [hunk 2]
```diff
@@ -4155,23 +4155,32 @@ VulkanReplayConsumerBase::OverrideCreateBuffer(PFN_vkCreateBuffer
...
+    uint32_t                              modified_queue_family_index = 0;
     // Check for a buffer device address.
     bool                uses_address         = false;
     VkBufferCreateFlags address_create_flags = 0;
     VkBufferUsageFlags  address_usage_flags  = 0;
+
+    if ((modified_create_info.sharingMode == VK_SHARING_MODE_CONCURRENT) &&
+        (modified_create_info.queueFamilyIndexCount > 0) && (modified_create_info.pQueueFamilyIndices != nullptr))
+    {
+        modified_create_info.queueFamilyIndexCount = 1;
```
Comment (andrew-lunarg) [inline framework/decode/vulkan_replay_consumer_base.cpp:4171]: I don't think we can assert that all buffers with concurrent usage only use the first queue family. This is a lie for traces in which buffers are used by multiple families and violates the Vulkan specification. 
https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkBufferCreateInfo

File: framework/decode/vulkan_replay_consumer_base.cpp [hunk 3]
```diff
@@ -4283,34 +4290,42 @@ VulkanReplayConsumerBase::OverrideCreateImage(PFN_vkCreateImage
...
+    auto                                  capture_id                  = (*pImage->GetPointer());
+    auto                                  replay_create_info          = pCreateInfo->GetPointer();
+    VkImageCreateInfo                     modified_create_info        = *replay_create_info;
+    uint32_t                              modified_queue_family_index = 0;
 
-    VkResult result = allocator->CreateImage(
-        pCreateInfo->GetPointer(), GetAllocationCallbacks(pAllocator), capture_id, replay_image, &allocator_data);
+    if ((modified_create_info.sharingMode == VK_SHARING_MODE_CONCURRENT) &&
+        (modified_create_info.queueFamilyIndexCount > 0) && (modified_create_info.pQueueFamilyIndices != nullptr))
+    {
```
Comment (andrew-lunarg) [inline framework/decode/vulkan_replay_consumer_base.cpp:4301]: Same point as for buffers.

**Review by andrew-lunarg — CHANGES_REQUESTED** [review]
I don't think we can just lie to the Vulkan implementation and say all these buffers and images are only used on queue family zero. It might help one particular troublesome trace on a particular implementation but it will break other traces and other implementations.

At a minimum, I think it should be hidden behind a flag and only turned on optionally, but even then it really doesn't make sense to me.

For traces with queue family trouble, I'd suggest implementing my issue https://github.com/LunarG/gfxreconstruct/issues/949 and then recapturing them so that they really do only use queue family zero.

**Comment by locke-lunarg** [comment]
@andrew-lunarg  You're right. I agreed with you.

**Pull request closed without merging**